### ISSUE TITLE
Add an option to dismiss notifications with ESC instead of minimizing them

### ISSUE THREAD

**Issue by bobbrow** [microsoft/vscode#149106]
I had always thought that pressing ESC had the same behavior as clicking the 'x' for the popup "toast" messages, but while investigating an issue with missing telemetry for an experiment we're doing in CMake Tools, I discovered that pressing ESC after calling `vscode.window.showErrorMessage` didn't return a result back to our code nor did it reject the promise.  The code just sat there until we shut down the extension host window.

This morning I had a thought that maybe the message had been minimized, and sure enough that's what it was.  This is a problem for extensions that want to log telemetry as a result of a dismissal.  Can we get an option in the API to make pressing ESC behave the same as clicking the 'x' button in certain toast messages?

**Comment by sana-ajani** [comment]
cc @digitarald

**Comment by digitarald** [comment]
I like that. When they go off screen the chances are low that anybody finds them again anyways. cc @daviddossett who has been thinking deeper about notification flows

**Comment by daviddossett** [comment]
I hear the concern—one worry I have here is associating ESC with the somewhat destructive "Clear Notification" action instead of minimizing/hiding it. There's an explicit action to clear them presumably to avoid accidentally missing the notification. That's something we've [heard complaints about before](https://twitter.com/idangazit/status/1503582722903609344).

<img width="405" alt="CleanShot 2022-05-12 at 16 16 57@2x" src="https://user-images.githubusercontent.com/25163139/168182494-bf992856-e1ee-450e-a232-fdd8a63f9952.png">

**Comment by daviddossett** [comment]
I'm not sure if there's another way to listen to a minimize event? @bpasero might know—however he's on paternity leave at the moment.

**Comment by bobbrow** [comment]
> one worry I have here is associating ESC with the somewhat destructive "Clear Notification" action instead of minimizing/hiding it.

I'm not asking to make it the default behavior.  I'm asking for an option in the API to flag a given notification as closeable when minimized.  These are the options today:

```ts
    export interface MessageOptions {

        /**
         * Indicates that this message should be modal.
         */
        modal?: boolean;
    }
```

It would be nice if a new property was added that could allow us to opt-in to the behavior for specific cases like this one where we need to collect telemetry.

**Comment by digitarald** [comment]
Would this also be solved if we fire an event when a notification gets hidden (aka, not closed or acted upon).

**Comment by daviddossett** [comment]
> Would this also be solved if we fire an event when a notification gets hidden (aka, not closed or acted upon).

That seems like a better solution here if the goal is to collect more accurate telemetry. At least when compared to having nondeterministic clear/close behavior depending on the notification.

**Comment by isidorn** [comment]
@bobbrow thanks for filling the issue. Great discussion.
I agree with @digitarald and @daviddossett. This is a cleaner solution for this use case instead of adding API. 

However with the current API the C++ extension can not do this. Since the `showMessage` API resolves a promise only once a user dismisses it or chooses a button. So there is not way for an extension to listen to when the user hides all notifications.

fyi @luabud might have ideas on how an extension can figure out if the user hid a notification for experiment purposes.

@bpasero might also have ideas once he returns (in around 2 weeks)

**Comment by bobbrow** [comment]
> Would this also be solved if we fire an event when a notification gets hidden (aka, not closed or acted upon).

We could make that work, but the programming model is more complicated because we may get future events (e.g. the await returns later, or maybe multiple minimizes?) if the user rehydrates the notification and reacts to it.  So we may have already sent an event in the session noting that the user minimized the notification, but we could end up sending another one if they choose one of the positive actions later.

To summarize, we'll take what we can get, but my first choice would be a new option to dismiss the notification when ESC is pressed.

**Comment by luabud** [comment]
Interesting! It didn't occur to me that hiding the notification doesn't return anything. I think it would be nice if extensions had an option to e.g. request a return for when the notification is hidden, or something like that. FWIW there is telemetry that differentiates when a notification is hidden (with ESC) vs when it's cleared (with the close button).

**Comment by digitarald** [comment]
I am concerned about changing the default as we don't have indicators that this is causing problems for users. We also have only hunches that users go back to hidden notifications, but even if few do it it is an escape hatch.

@bobbrow in your case, is the notification `sticky`? Maybe that could be a special case – that sticky notifications are easier to dismiss. One could argue that sticky notifications act more like modals that can not be re-hydrated.

**Comment by bobbrow** [comment]
> I am concerned about changing the default

Hopefully I was clear earlier that I'm not asking you to change the default.  I'm asking you to add an option that we can opt-in to that changes the behavior of the ESC key press.

> @bobbrow in your case, is the notification `sticky`?

I don't know what you mean by `sticky`.  We call this API:
```ts
showErrorMessage(message: string, options: MessageOptions, ...items: string[]);
```
which has no "sticky" option.  It has a "modal" option, but we don't want to use that because we don't want to be an annoyance.

**Comment by luabud** [comment]
I think we all agree with not changing the default behaviour, it's more about providing an option to extensions to be able to do something when users press ESC. 
For example, I have a custom extension with the following code:

```
 const selection = await window.showErrorMessage(
            'It looks like your Python project has not been set up yet. Would you like us to create a virtual environment and install the dependencies?',
            'Yes, set up project',
            'No, thanks',
        );
 ```

When users close the notification, `selection` gets `undefined`. But if they press ESC, nothing ever happens (and I think this is what @bobbrow meant by "the code just sat there until we shut down the extension host window.")

**Comment by bpasero** [comment]
Being able to press `ESC` was meant as a way to quickly ignore the notification and go back to coding but with being able to restore it later from the notifications center. I would find it strange as a user to sometimes see a notification dismiss and sometimes hide without knowing exactly why that is. 

To indicate the notification is still pending we show a dot:

<img width="81" alt="image" src="https://user-images.githubusercontent.com/900690/170645543-3d6c649a-1020-438b-b3c0-f0adb349492c.png">

If it is mission critical for the extension to get attention from the user, a modal dialog should rather be used imho.

**Comment by bobbrow** [comment]
It is not mission critical to respond to the dialog and we do not want to annoy our users, so I think a model dialog is the wrong choice for this case.  But I think extension authors should be allowed to control the behavior. Is there data available that can show us how frequently users actually go back to the minimized notifications?

My intuition - based on how long we had to debug the extension to understand the behavior - is that there may be a significant cohort of users who do not believe that pressing ESC minimizes notifications, instead they believe that it should be dismissing them.  But I am not advocating for a change in the default behavior to match my bias.  I am advocating for a simple addition to the API that will allow extension authors to choose what happens when the notification is minimized.  The current behavior complicates our ability to create A/B experiments that interact with notifications.

![image](https://user-images.githubusercontent.com/12818240/170736320-64679c99-f02f-4f58-8bbc-54ba5095c0b8.png)

Can you please elaborate on what additional information you need?

**Comment by digitarald** [comment]
@bobbrow there will always be unknowns in users dismissing/ignoring the prompt (close or reload while VS Code, future explorations of do-not-disturb mode). For a/b experimenting I recommend the click-through-rate of how many prompts are shown vs how many users take the intended action. Did you consider this as an option and have found downsides?

Not dismissing on Esc doesn't serve many, but the few that hit Esc unintentionally and need a way to get back to those notifications.

**Comment by bobbrow** [comment]
> Not dismissing on Esc doesn't serve many, but the few that hit Esc unintentionally and need a way to get back to those notifications.

There are cases where we don't care if they can get back to the notification.  Not all notifications are equal in importance.  We just want the promise to complete, but instead it hangs forever.

> I fear extensions could abuse this to just bring up the same notification again and spam the user.

Extensions that abuse this will find that their users will quickly uninstall them.  I don't believe this is a solid argument against having the feature.

> For a/b experimenting I recommend the click-through-rate of how many prompts are shown vs how many users take the intended action. Did you consider this as an option and have found downsides?

We haven't tried it yet, but a downside for starters is that it requires two events instead of one.  My PM will have to look into how this affects the experiments, but she is on vacation right now.  At the very least I assume she'll have to do more work to compute the results given that it would not be just a single event that we'd be tracking anymore.  You have to correlate events between sessions and stuff to make sure that you're not pairing the telemetry where you got a response against non-existent telemetry because the notification was minimized and the 2nd event didn't fire.

**Comment by isidorn** [comment]
I do not see any actions planned from the above discussion. Thus I am closing this.
If there are good action item suggestions do let me know and I can reopen
